### ISSUE TITLE
using provider with modules

### ISSUE THREAD

**Issue by jseaidou** [hashicorp/terraform#6132]
We are trying to modularize our configuration files and we are running into a issue where the `provider` field can't find a provider in a different module. 

we have a directory of the following structure:

```
common
    |__ modules.tf
    |__ providers
          |__ openstack.tf
modules.tf
project
    |__ main.tf
```

The root level `modules.tf` looks like this: 

```
module "common" {
    source = "./common"
}

module "project" {
    source = "./dev"
}
```

similarly, `modules.tf` in `common/` looks like this

```
module "providers" {
    source = "./providers"
}
```

`openstack.tf` is very straightforward 

```
provider "openstack" {
    [config]
    alias = "my_alias"
}
```

finally in `main.tf` we are doing the following:

```
resource "openstack_compute_instance_v2" "test_instance" {
    [config]
    provider = "module.common.providers.openstack.my_alias"
}

```

for `provider` field, we have also tried the following combination:

```
provider = "module.common.module.providers.openstack.my_alias"
provider = "common.providers.openstack.my_alias"
provider = "${module.common.module.providers.openstack.my_alias}"
provider = "${module.common.providers.openstack.my_alias}"
provider = "${module.common.module.providers}.openstack.my_alias"
provider = "${module.common.providers}.openstack.my_alias"
```

It seems like we can't have providers outside of the same module, but we have certain providers that we'd like to use across modules and don't want to just dump in top level but rather in their own module

my current version is:

```
> terraform version
Terraform v0.6.15-dev (3a8e7d4b651ec71cc7cea821ed0df4a4bd7a3f26)
```

this seems misleading however since I am actually using branch `dev-0.7` and compiled the binary from there.

**Comment by agalazis** [comment]
any answer?

**Comment by apparentlymart** [comment]
Hi all! Sorry for the long silence here.

At the time this issue was filed the support for sharing providers between modules was very weak, but in Terraform 0.11 we added features that allow for providers to be explicitly passed between modules. [The v0.11 upgrade guide](https://www.terraform.io/upgrade-guides/0-11.html#interactions-between-providers-and-modules) discusses what was changed and links to other documentation.
